### PR TITLE
Fix numerical stability in binomial.log_prob

### PR DESCRIPTION
**Pull request by neerajprad**
This issue was discovered by @fehiepsi in https://github.com/uber/pyro/issues/1706 with the `log_prob` computation for Binomial, ~and can be seen with `torch.float32` when we have a combination of low probability value and high `total_count` - a test is added to capture this (since scipy only uses float64, the comparison is done using relative tolerance).~ 

The problem is in the code that tries to pull out the minimum values amongst the logits (written by me earlier, presumably to avoid numerical instability issues), but it is not needed.

EDIT: After a few attempts, I have been unable to reliably show that the change is more numerically stable, and have removed my previous test which fails on linux. The reason is that the issue manifests itself when `total_count` is high and `probs` is very low. However, the precision of `lgamma` when `total_count` is high is bad enough to wash away any benefits. The justification for this still stands though - (a) simplifies code (removes the unnecessary bit), (b) is no worse than the previous implementation, (c) has better continuity behavior as observed by @fehiepsi in the issue above.

cc. @fehiepsi, @alicanb, @fritzo

### CODE REVIEW

File: torch/distributions/binomial.py [hunk 1]
```diff
@@ -113,11 +113,9 @@ def log_prob(self, value):
         log_factorial_n = torch.lgamma(self.total_count + 1)
         log_factorial_k = torch.lgamma(value + 1)
         log_factorial_nmk = torch.lgamma(self.total_count - value + 1)
-        max_val = (-self.logits).clamp(min=0.0)
-        # Note that: torch.log1p(-self.probs)) = max_val - torch.log1p((self.logits + 2 * max_val).exp()))
+        # Note that: torch.log1p(-self.probs)) = - torch.log1p(self.logits.exp()))
         return (log_factorial_n - log_factorial_k - log_factorial_nmk +
-                value * self.logits + self.total_count * max_val -
-                self.total_count * torch.log1p((self.logits + 2 * max_val).exp()))
+                value * self.logits - self.total_count * torch.log1p(self.logits.exp()))
```
Comment (fehiepsi) [inline torch/distributions/binomial.py]: In PyTorch, `logits != probs.exp()`, so this formula seems wrong to me. Previously, the tests passed, which makes me confused.

Comment (neerajprad) [reply torch/distributions/binomial.py]: @fehiepsi - That is correct - `logits != probs.log()` for the binary case. 

Does this seem reasonable?
```
value * probs.log() + (total_count - value) * torch.log1p(-probs))
 = value * (probs.log() - torch.log1p(-probs)) + total_count * torch.log1p(-probs)
 = value * logits + total_count * (1 - probs)
 = value * logits - total_count * torch.log1p(self.logits.exp()))
```

Comment (fritzo) [reply torch/distributions/binomial.py]: first equation: correct
second equation: correct
third equation: incorrectly assumes `probs=logits.exp()`

Comment (fehiepsi) [reply torch/distributions/binomial.py]: Thanks! It seems reasonable to me now. @fritzo will be confused. :D

Comment (neerajprad) [reply torch/distributions/binomial.py]: Just in case, this causes confusion for others, expanding on the last step:

```
(1 - probs).log()
 = (1 - (logits.exp()/(1 + logits.exp())).log()
 = (1 / (1 + logits.exp())).log()
 = - (1 + logits.exp())).log()
 = - log1p(logits.exp())
```

**Comment by neerajprad** [comment]
`torch.lgamma` seems to have precision issues with FloatTensor in linux (much more so than OSX) due to which it isn't passing the test that I added. I will modify the test accordingly.

**Review by ezyang — APPROVED** [review]
